File: Core-backend-layer/Programming-fundamentals/Java/OOP/src/AbstractClasses/products/GroceryProduct.java
package products;

public class GroceryProduct extends Product {
    int quantityInStock;

    public GroceryProduct(int productId, String name, double price, double costPrice, int quantityInStock) {
        super(productId, name, costPrice);
        this.price = price; // not initialized in the constructor of the abstract class
        this.quantityInStock = quantityInStock;
    }

    // Implementation of abstract method
    @Override
    public void displayInfo() {
        System.out.println("Grocery product: " + name + ", Price: $" + price +
                ", Quantity in stock: " + quantityInStock + ", Tax: $" + calculateTax() +
                ", Profit: $" + calculateProfit());
    }

    // Method using protected field from superclass
    public void orderMoreStock(int additionalQuantity) {
        quantityInStock += additionalQuantity;
        System.out.println("Ordered " + additionalQuantity + " more units of " + name + " Product ID: " + productId + ". New stock: " + quantityInStock);
    }
}

File: Core-backend-layer/Programming-fundamentals/Python/OOP/src/abstract_classes/products/grocery_product.py
from products.product import Product

class GroceryProduct(Product):
    def __init__(self, product_id, name, price, cost_price, quantity_in_stock):
        super().__init__(product_id, name, cost_price)
        self.price = price  # initialized in the constructor of the abstract class to 0
        self.quantity_in_stock = quantity_in_stock

    # Implementation of abstract method
    def display_info(self):
        print(f"Grocery product: {self.name}, Price: ${self.price}, "
              f"Quantity in stock: {self.quantity_in_stock}, Tax: ${self.calculate_tax()}, "
              f"Profit: ${self._calculate_profit()}")

    # Method using internal intended field from superclass
    def order_more_stock(self, additional_quantity):
        self.quantity_in_stock += additional_quantity
        print(f"Ordered {additional_quantity} more units of {self.name} Product ID: {self._product_id}. New stock: {self.quantity_in_stock}")